Build/distributable (#1297)
o3de/o3de.org: content/docs/user-guide/build/distributable-engine.md

1.  Generate the toolchain project files using a unique `LY_VERSION_ENGINE_NAME` CMake cache setting. This value is the name of the engine used by the O3DE project manager and registration system.
    
    ```cmd
    cmake -B build/windows_vs2019 -G "Visual Studio 16" --config profile -DLY_VERSION_ENGINE_NAME="MyO3DE"
    ```

    The **profile** configuration is recommended for distributed builds, as it provides additional logging and engine introspection capabilities useful during project development at minimal performance cost. **debug** builds are primarily useful during engine development.
1.  Build the `INSTALL` target.
    
    ```cmd
    cmake --build build/windows_vs2019 --target INSTALL --config profile
    ```
    
    The binaries will be placed into the directory specified by the `CMAKE_INSTALL_PREFIX` CMake cache variable, or in an `install` subdirectory of the source code by default.

Run the following steps from the install directory:

1. Install the version of Python and modules required by the engine.
   
   ```cmd
   python\get_python.bat
   ```

### Register the engine

> The next step in creating a distributable build of the engine tools is to register the locally installed engine. These steps should be re-run every time you create a new build of the `INSTALL` target.

Reviewer: ```suggestion
The next step in creating a distributable build of the engine tools is to register the locally installed engine. You must re-run these steps every time you create a new build of the `INSTALL` target.
```

> The next step in creating a distributable build of the engine tools is to register the locally installed engine. These steps should be re-run every time you create a new build of the `INSTALL` target.

Reviewer 2: I'm not sure why you'd need to re-register the engine every time you build it to the install target, assuming the engine name and install path are kept the same. I'm curious as to the reason for stating this requirement.

Reviewer 3: The engine only need to be registered if the `LY_VERSION_ENGINE_NAME` value was changed as that effects the engine name in the install layout

> The next step in creating a distributable build of the engine tools is to register the locally installed engine. You must re-run these steps every time you create a new build of the `INSTALL` target.

Reviewer 2: ```suggestion
The next step in creating a distributable build of the engine tools is to register the locally installed engine. You must re-run this registration step any time you make a change to the CMake cache variable `LY_VERSION_ENGINE_NAME` or to the `engine_name` in `engine.json`.
```

Reviewer 2: @lumberyard-employee-dm How's this for a rewording?

Reviewer 3: Could we clarify that the engine.json is located at `<engine-root>/engine.json` or something that points the user to the engine source root directory?
Users might not associate the `engine.json` file with being inside of the engine root immediately

Besides that, the rewording of that section is good.

Reviewer 2: Actually, to avoid the potential for confusion here between the source `engine.json` and the installed `engine.json`, I'm going to remove the mention of `engine.json` in my rewording, especially since these instructions have been using `LY_VERSION_ENGINE_NAME`.

Register the engine locally with the `o3de` script. Run the following command from your distributable build directory:
    
```cmd
scripts\o3de.bat register --this-engine
```

{{< note >}}
If there is a conflict in the existing `o3de_manifest.json` file on the system, the engine registration will fail and the `o3de` output provides information on how to fix the error. We recommend either automating this process through a script (for teams with rapid iteration requiring new builds) or using unique engine names for each new distributable (for teams requiring stability, rollback to previous builds outside of source control, or with infrequent distributable updates).
{{< /note >}}

At this point, the installed engine build is now registered in the O3DE config directory (by default `%USERPROFILE%\.o3de`) in the `o3de_manifest.json` file.

## Project association

At this point, you're ready to either create a new project for distribution or re-associate an existing project with the install build that will be used to create the distributable.

### Create a new project

> Run the following command from the local engine source directory:

Reviewer 2: Wait, shouldn't the new project be created from the distributable build directory, not the engine source directory?

Reviewer 3: It should be created from the distributable install directory.

Using the engine source directory would result in the same project being created, BUT with the project being registered using the source engine name and not the SDK layout engine name which was set via the `LY_ENGINE_VERSION_NAME`

Reviewer 2: Yes! That's exactly what I was worried about. Thanks for verifying. Will update.

                    "LoadLevel": "MyLevel"
                }
            }
        }
    }
    ```

To verify that the level loads when the game is launched, run the project's game launcher:

```cmd
W:\MyProject\build\windows_vs2019\bin\profile\MyProject.GameLauncher.exe
```

## Create a distributable build

The final step to create the binaries for distribution to project developers is to create the set of install binaries that they can use without building the engine from source. Like you did to create the local binaries for the engine, you'll now create binaries for the project.

From the project directory, run the following steps:

1. Create an install build of the project.
    
    ```cmd
    cmake --build build\windows_vs2019 --config profile --target INSTALL
    ```

    This creates the binaries in an `install` subdirectory of your project. For our example it would be `W:\MyProject\install`.

## Distribute your build

At this point, the created `install` directory is your final set of distributables. It's up to your team to decide where to host the build, but we recommend a source control system that has good blob support.

{{< important >}}

> The .gitignore rule automatically adds the following folders to ignore for git. If using another source control system, make sure the following subdirectories of your project are ignored and not checked in:

Reviewer 3: It may be good to notate that the .gitignore file is located in the project root after project creation and can be referenced to determine the list of directories/files  to ignore

> The .gitignore rule automatically adds the following folders to ignore for git. If using another source control system, make sure the following subdirectories of your project are ignored and not checked in:

Reviewer 2: ```suggestion
The .gitignore rule automatically adds the following folders to ignore for git. The `.gitignore` file is located in the project root after project creation. If using another source control system, make sure the following subdirectories of your project are ignored and not checked in:
```

    
    ```
    \user
    \build
    \Cache
    ```

> These files are generated locally during builds, and checking them in may cause issues for developers.

Reviewer 2: ```suggestion
These files are generated locally during builds, and checking them in may cause issues for developers. Reference the `.gitignore` file in the project root to verify the list of directories and files to ignore.
```

{{< /important >}}

At this point, you're ready to have your team perform local setup for either engine development or project development.

### Engine development setup

Engine developers will work on O3DE source code and create new distributable builds. To enable their workflows, perform the following steps on their workstations:

1. Get your modified engine source code from its hosted location.
1. Install the Python runtime and modules required by the engine. Run the following command from the directory containing engine source:

    ```cmd
    python\get_python.bat
    ```
1. Register the engine locally. Run the following command from the directory containing engine source:

    ```cmd
    scripts\o3de.bat register --this-engine
    ```
1. Get the project from its hosted location.
1. Register the project with the _local development engine_. Run the following command from the directory containing the engine source:
   
   ```cmd
   scripts\o3de.bat register -pp W:\MyProject
   ```
1. Follow the [project build instructions](https://o3de.org/docs/user-guide/build/) to generate any project-specific builds, such as the launcher.

### Project developer setup

> Project developers will work on project-specific source code or use O3DE creative tools to add project content. To enable their workflows, project developers are required to _only_ pull a project from its hosted location.

Reviewer 2: I suggest rewording this sentence. I'm not 100% sure what is really meant here. The emphasis of the word _only_, together with its location in the sentence, is awkward. It seems to be trying to convey that project developers _must_ pull a project from its hosted location, and no other location. If this is truly the case, then I'd start by replacing "project developers are required to only" with "project developers must only". Then, at the beginning of that sentence, in place of "To enable their workflows", I'd supply a more precise reason why they must do this.

Reviewer 3: I hesitate calling this "Project developer setup".

This seems like information that is in addition to the engine of the "Engine development setup" section.
For users that don't need to re-build the project, they can use the prebuild binaries in the `<project-root>/install/bin/Windows/profile/Default` subdirectory.

Reviewer 2: Leaving this for @sptramer to address in a follow up.
